File: scripts/ingest.ts
/**
 * Bulk knowledge ingestion CLI.
 *
 *   npm run ingest -- --persona <slug-or-id> <file-or-dir> [more files...]
 *
 * Examples:
 *   npm run ingest -- --persona eugene-schwartz ./books/breakthrough.pdf
 *   npm run ingest -- --persona eugene-schwartz ./knowledge   (all files in a dir)
 *
 * Reads OPENAI_* and DATABASE_URL from .env.local (same as the app).
 */
import { readFileSync, readdirSync, statSync } from 'node:fs'
import { join, basename } from 'node:path'
import { loadEnvConfig } from '@next/env'

// Load env BEFORE importing modules that read it.
loadEnvConfig(process.cwd())

async function main() {
  const { queryOne, getPool } = await import('../lib/db')
  const { ingestText, extractText } = await import('../lib/ingest')

  const args = process.argv.slice(2)
  const personaIdx = args.indexOf('--persona')
  if (personaIdx === -1 || !args[personaIdx + 1]) {
    console.error('Uso: npm run ingest -- --persona <slug-ou-id> <arquivo-ou-pasta> [...]')
    process.exit(1)
  }
  const personaRef = args[personaIdx + 1]
  const paths = args.filter((a, i) => i !== personaIdx && i !== personaIdx + 1 && !a.startsWith('--'))
  if (paths.length === 0) {
    console.error('Nenhum arquivo/pasta informado.')
    process.exit(1)
  }

  // Resolve persona by slug or id (id cast guarded so a slug doesn't break ::uuid).
  const persona = await queryOne<{ id: string; name: string; slug: string }>(
    `select id, name, slug from personas
      where slug = $1 or id::text = $1
      limit 1`,
    [personaRef]
  )

  if (!persona) {
    console.error(`Persona não encontrada: ${personaRef}`)
    process.exit(1)
  }
  console.log(`→ Persona: ${persona.name} (${persona.slug})`)

  // Expand directories into their files.
  const files: string[] = []
  for (const p of paths) {
    if (statSync(p).isDirectory()) {
      for (const f of readdirSync(p)) files.push(join(p, f))
    } else {
      files.push(p)
    }
  }

  for (const file of files) {
    try {
      const buffer = readFileSync(file)
      const text = await extractText(buffer, basename(file))
      if (!text.trim()) {
        console.warn(`  ⚠ ${basename(file)}: sem texto extraível, pulado.`)
        continue
      }
      const result = await ingestText(persona.id, text, basename(file))
      console.log(`  ✓ ${basename(file)}: ${result.chunks} trechos indexados.`)
    } catch (err) {
      console.error(`  ✗ ${basename(file)}: ${(err as Error).message}`)
    }
  }

  console.log('Concluído.')
  await getPool().end()
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: backend/app/cli/ingest.py
"""Bulk knowledge ingestion CLI. Port of scripts/ingest.ts.

  python -m app.cli.ingest --persona <slug-or-id> <file-or-dir> [more...]

Reads OPENAI_* and DATABASE_URL from the environment (or ../.env.local).
"""
from __future__ import annotations

import asyncio
from pathlib import Path

import typer

app = typer.Typer(add_completion=False)


async def _run(persona_ref: str, paths: list[str]) -> None:
    from ..db import close_pool, query_one
    from ..services.ingest import extract_text, ingest_text

    persona = await query_one(
        "select id, name, slug from personas where slug = $1 or id::text = $1 limit 1",
        [persona_ref],
    )
    if not persona:
        typer.echo(f"Persona não encontrada: {persona_ref}", err=True)
        raise typer.Exit(1)
    typer.echo(f"→ Persona: {persona['name']} ({persona['slug']})")

    # Expand directories into their files.
    files: list[Path] = []
    for p in paths:
        path = Path(p)
        if path.is_dir():
            files.extend(sorted(path.iterdir()))
        else:
            files.append(path)

    for file in files:
        try:
            text = extract_text(file.read_bytes(), file.name)
            if not text.strip():
                typer.echo(f"  ⚠ {file.name}: sem texto extraível, pulado.")
                continue
            result = await ingest_text(str(persona["id"]), text, file.name)
            typer.echo(f"  ✓ {file.name}: {result['chunks']} trechos indexados.")
        except Exception as err:  # noqa: BLE001
            typer.echo(f"  ✗ {file.name}: {err}", err=True)

    typer.echo("Concluído.")
    await close_pool()


@app.command()
def main(
    paths: list[str] = typer.Argument(..., help="Arquivos ou pastas para indexar."),
    persona: str = typer.Option(..., "--persona", help="slug ou id da persona."),
) -> None:
    asyncio.run(_run(persona, paths))


if __name__ == "__main__":
    app()
